[1단계 - 음식점 목록] - 데이지(김유경) 미션 제출합니다.

kimyou1102 opened the pull request:
안녕하세요 하리! 이번 미션은 니야🐢와 함께 페어를 진행했습니다. 컴포넌트 구조를 고민하며 진행해봤는데 이번 미션 잘부탁드립니다!! 

## 학습 목표

이번 미션을 통해 다음과 같은 학습 경험들을 쌓는 것을 목표로 합니다.

- UI와 도메인 영역을 분리할 수 있는 설계를 고민해보고, 목적에 맞게 객체와 함수를 활용
- TDD 방식으로 개발하며, 단위 테스트 기반으로 점진적인 리팩터링

## 제출 전 체크 리스트

<!-- 리뷰를 요청하기 전, 다음 항목들을 기본적으로 확인해 주세요. -->

- [x] 기능 요구 사항을 모두 구현했고, 정상적으로 동작하는지 확인했나요?
- [x] 기본적인 프로그래밍 요구 사항을 준수하고 있는지 확인했나요?
- [x] 테스트 코드는 모두 정상적으로 실행되나요?
- [x] (해당하는 경우) 배포한 데모 페이지에 정상적으로 접근할 수 있나요?
  - 배포한 링크 기입: https://kimyou1102.github.io/javascript-lunch/

## 리뷰 요청 & 논의하고 싶은 내용

<!-- PR 작성자로서, 코드만으로는 알기 어려운 작성자의 의도와 문제 해결 과정에 대해 공유해 주세요.
이 PR에서 달성해야 하는 학습 목표를 잘 달성하고 있는지 스스로 확인하며, 피드백 받을 수 있는 내용일지 점검해 보세요.
리뷰어에게 정답을 묻기보다 고민하고 의사 결정한 과정에 대해 공유하고 이에 대한 피드백을 받으며 대화해보기를 권장합니다. -->

### 1) 이번 단계에서 가장 많이 고민했던 문제

<!-- 구현 과정에서 가장 어려웠던 점이나 많이 고민한 점은 무엇인가요?
이를 해결하기 위해 어떤 방법들을 검토하고 시도했으며, 그 과정에서 새롭게 배운 점이 있나요? -->

**1. 명세 변경에 유연한 컴포넌트 작성법**
ButtonGroup을 하나의 컴포넌트로 묶으려고 했는데 명세가 변경될 때마다 `ButtonGroup`의 `props`가 계속 늘어나게 될 가능성이 있습니다. (예를들어 type이 다르다던가, style이 달라서 className을 추가해야 한다던가 등의 추가적인 상황을 생각했습니다.)
    
그렇게 `props` 가 많아지면 코드가 복잡해지는 거 같아서 ButtonGroup을 묶지 않는 것을 선택했습니다. 재사용성과 유지보수성을 고려하면서도 이 두 개의 버튼을 효과적으로 묶을 방법이 있을지 고민입니다. 현재 방식이 괜찮을지 혹은 더 좋은 방법이 있다면 말씀해주시면 감사하겠습니다!
    

```
export default function ButtonGroup({
  text: { left, right },
  onClick: { onLeftClick = () => {}, onRightClick = () => {} },
}) {
  const $buttonContainer = document.createElement("div");
  $buttonContainer.className = "button-container";

  const $leftButton = Button({ text: TEXT_TYPES[action], action: action });
  const $rightButton = Button({
    type: "submit",
    text: "추가하기",
    action: "add",
  });

  $buttonContainer.appendChild($leftButton);
  $buttonContainer.appendChild($addButton);

  $leftButton.addEventListener(EVENT_TYPES.click, onRightClick);

  return $buttonContainer;
}

```

**2. 추상화 레벨 맞추기 vs 접근성**
`SelectBox`를 활용한 DistanceSelect 컴포넌트를 만들었는데 추상화 레벨을 맞추려다 보니 고민이 생겼습니다. 외부에서 보면 추상화 레벨이 통일되어 가독성이 좋지만 내부에서는 추가적인 경로를 거치는 느낌이 들어서 불필요한 추상화 같기도 합니다. 추상화 레벨 맞추기 vs 접근성을 높이기 중 어떤 것을 더 신경 써야 할까요? 혹은 둘 다 만족할 수 있는 방법이 있을까요??

```
export default class DistanceSelect {
  render() {
    const $distanceSelect = new SelectBox({
      label: "distance",
      options: DISTANCE,
    }).render();

    return $distanceSelect;
  }
}
```

### 2) 이번 리뷰를 통해 논의하고 싶은 부분

<!-- 구현한 코드와 학습 목표와 관련해 피드백을 받고 싶은 부분이나, 함께 논의해보고 싶은 점 -->

- Cypress에서 When-Then-Given 같은 가독성을 높이는 작성 방식이 있을까요? 현재 저희는 주석 없이 공통적인 부분을 함수로 추출한 후 그다음 시나리오를 테스트하는 방식으로 진행하고 있습니다. 이보다 더 가독성을 높일 방법이 있을지 하리의 의견이 궁금합니다!
- E2E 테스트를 어디까지 커버해야 할지 고민입니다. 현재는 버튼을 클릭 → 모달오픈 → 폼 작성 → 리스트에 추가 확인까지 테스트했습니다. 한과정을 끝까지 테스트 하는 것이라 생각되어 이렇게 작성했는데 여러 개 제출을 확인하는 등 더 세부적인 테스트도 포함해야 할지 고민됩니다. E2E 테스트의 범위를 어떻게 설정하는 것이 효과적일지 궁금합니다!
- 컴포넌트의 응집성을 높이기 위해 각 컴포넌트마다 개별 폴더를 만들었습니다. 저희는 통일성을 위해 CSS 파일이 없는 컴포넌트도 개별 폴더를 두었는데 보다 보니 통일성은 유지되지만 오히려 가독성이 떨어지는 느낌이 듭니다. CSS가 없는 컴포넌트도 폴더를 유지하는 것이 좋을지 혹은 더 나은 구조가 있을지 하리의 의견이 궁금합니다!

---

<details>
<summary>✅ 리뷰어 체크 포인트</summary>
<!-- 리뷰어가 이 PR을 검토할 때 중점적으로 확인할 사항입니다.-->

### 1단계

- [ ] UI를 컴포넌트 단위로 분리했는가?
- [ ] 재사용 가능한 컴포넌트를 고려했는가?
- [ ] 주요 기능 시나리오에 대한 E2E 테스트를 작성했는가?

### 2단계

- [ ] 오류 상황(예: 잘못된 입력)에 대해서도 테스트 케이스를 정의했는가?
- [ ] 타입 정의를 적절히 사용하여 컴파일 오류 없이 안정적으로 동작하는가?
- [ ] 로컬/배포 환경에서 테스트가 정상적으로 동작하는지 확인했는가?

</details>

sooyeoniya committed:
docs: README.md 기능 요구사항 및 구현 목록 작성

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

sooyeoniya committed:
feat(Header): Header 컴포넌트 생성

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

sooyeoniya committed:
chore: 이미지 폴더 위치 변경

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

sooyeoniya committed:
feat(RestaurantListItem): RestaurantListItem 컴포넌트 생성

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

sooyeoniya committed:
feat(RestaurantList): RestaurantList 컴포넌트 생성

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

sooyeoniya committed:
feat(SelectBox): SelectBox 컴포넌트 생성

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

sooyeoniya committed:
feat(BottomSheet): BottomSheet 컴포넌트 생성

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

sooyeoniya committed:
chore: html, css 파일 위치 이동 및 html 태그 수정

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

sooyeoniya committed:
feat(Button): Button 컴포넌트 생성 밎 BottomSheet에 적용

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

sooyeoniya committed:
feat(DescriptionInput): DescriptionInput 컴포넌트 생성

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

sooyeoniya committed:
feat(RestaurantNameInput): RestaurantNameInput 컴포넌트 생성

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

sooyeoniya committed:
feat(LinkInput): LinkInput 컴포넌트 생성

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

sooyeoniya committed:
refactor(BottomSheet): BottomSheet 내부에 분리된 컴포넌트 적용

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

sooyeoniya committed:
feat(main): 전체 레이아웃 구성

- 구현한 내부 컴포넌트 가져와서 레이아웃 그리기
Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

sooyeoniya committed:
chore: 음식점 리스트 처음에 빈 값으로 초기화

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

sooyeoniya committed:
feat(Header): 새로운 음식점 추가 버튼 클릭 시 모달 창 열기

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

sooyeoniya committed:
feat(BottomSheet): 취소하기 버튼 및 백드롭 클릭 시 모달 닫기

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

sooyeoniya committed:
refactor(SelectBox): label 상수화 및 적용

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

sooyeoniya committed:
feat(BottomSheet): 새로운 음식점 추가 기능 구현

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

sooyeoniya committed:
feat(constants): 카테고리 및 거리 배열 상수화 및 적용

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

sooyeoniya committed:
docs: README.md 리팩토링 및 테스트 목록 작성

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

sooyeoniya committed:
refactor(Button): 버튼 스타일 variant 분기 처리 리팩토링

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

sooyeoniya committed:
refactor(BottomSheet): BottomSheet 컴포넌트 분리 및 SelectBox 폴더 위치 수정

- BottomSheet의 틀과 내부 폼 데이터를 각각 BottomSheetBase와 RestaurantForm으로 분리하여 폼을 주입하도록 함

- SelectBox는 common 폴더로 옮김

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

sooyeoniya committed:
refactor(BottomSheet): 카테고리 및 거리 선택 박스 컴포넌트 분리

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

sooyeoniya committed:
chore: 사용하지 않는 import 제거

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

sooyeoniya committed:
refactor(Button): Button 컴포넌트 common 폴더로 경로 수정

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

sooyeoniya committed:
refactor(RestaurantListSection): 레스토랑 리스트 및 아이템 폴더 경로 수정

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

sooyeoniya committed:
refactor: BottomSheetBase 경로 common 폴더로 옮김

- BottomSheetBase는 공통으로 사용될 수 있는 바텀시트 틀이라고 판단하여 common 폴더로 옮김

- 그 과정에서 BottomSheetBase를 사용하는 RestaurantForm 관련 컴포넌트도 RestaurantFormSection 으로 폴더 경로 수정

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

sooyeoniya committed:
refactor(CSS): 응집성 향상을 위한 CSS 파일 분리 및 이동

- 각 컴포넌트 폴더에 위치하도록 함

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

sooyeoniya committed:
refactor(constants): 이벤트 타입 상수화 및 적용

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

sooyeoniya committed:
refactor(App): 루트 JS 파일을 App으로 변경

- main보다 App이라는 명칭을 사용해 파스칼 케이스로 컴포넌트 느낌을 주어서 함수로 묶어서 실행하도록 함

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

sooyeoniya committed:
refactor(constants): 각 label 명 상수화 및 적용

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

sooyeoniya committed:
fix(SelectBox): distance label textContent 수정

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

sooyeoniya committed:
test: 새로운 음식점 추가 플로우 E2E 테스트 구현

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

sooyeoniya committed:
test: 바텀 시트 닫는 기능 E2E 테스트 구현

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

sooyeoniya committed:
docs: README.md 컴포넌트 폴더 구조 추가

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

sooyeoniya committed:
refactor(constants): 버튼 텍스트 상수화

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

sooyeoniya committed:
refactor: 전체 클래스 구조로 변경

- 함수 내부가 길어졌기 때문
- 렌더링과 이벤트 핸들링 로직 분리 및 해당 메서드들 재사용하기 위함

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

sooyeoniya committed:
refactor(constants): 버튼 타입 상수화 및 적용

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

sooyeoniya committed:
refactor: 이벤트 핸들러 함수 분리

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

sooyeoniya committed:
chore: 사용하지 않는 import 제거

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

sooyeoniya committed:
fix(BottomSheetBase): form 태그 중복 삭제

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

sooyeoniya committed:
refactor: App에서 상태관리 되도록 구조 변경

- App에서 상태를 관리하고, 이를 하위 컴포넌트로 전달하여 중앙에서만 상태관리하도록 수정
- 2단계 작업에 대한 확장성을 고려하여 상태 변경을 일관된 방식으로 관리해줌

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

sooyeoniya committed:
test: 필수 정보 입력하지 않았을 경우 음식점 추가되지 않는 플로우 테스트

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

sooyeoniya committed:
chore: RestaurantNameInput 앞에 Restaurant 제거

- 다른 입력 박스들과 통일해주기 위함

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

kimyou1102 committed:
refactor: 이미지 경로 수정

kimyou1102 committed:
chore: 배포 homepage 주소 변경

LAH1203 commented on README.md:
````diff
@@ -0,0 +1,83 @@
+```
+components
+├── 📂 common
+│ ├── 📂 bottomSheetBase
````
구현해주신 컴포넌트 파일은 파스칼 케이스인 반면, 폴더명은 카멜 케이스네요. 이유가 있을까요?

kimyou1102 replied:
컴포넌트명은 대문자로 시작해서 폴더명은 컴포넌트 파일과 구분하기 위해서 카멜케이스로 했습니다!

LAH1203 commented on README.md:
```diff
@@ -0,0 +1,83 @@
+│ │ └── button.css
+│ └── 📂 selectBox
+│   └── SelectBox.js
+├── 📂 header
```
헤더는 common 밖에서 만들어주신 이유가 있나요?

kimyou1102 replied:
layout폴더 안에 넣을까 고민했지만 폴더 안에 넣을 컴포넌트가 header하나밖에 없었습니다. 그래서 추후에 layout 컴포넌트가 더 생기면 이동시키자는 결론을 내서 components폴더 아래에 뒀습니다!

LAH1203 commented on index.html:
```diff
@@ -1,13 +1,13 @@
-    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
+    <meta charset="utf-8" />
     <title>점심 뭐 먹지</title>
+    <link rel="stylesheet" href="style.css" />
+    <link rel="icon" type="image/svg+xml" href="/vite.svg" />
+    <meta name="viewport" content="width=device-width, initial-scale=1" />
```
1. meta 태그를 아래로 옮기고 link 태그를 위로 올려주셨어요. 이렇게 하셨을 때 기대하신 동작이 있을까요?
2. `<meta name="viewport" content="width=device-width, initial-scale=1" />`는 어떤 동작을 할까요?

kimyou1102 replied:
1. 미션에서 제공된 기본 link태그를 지우고 미션을 진행하며 만든 style.css를 추가했습니다. 보통 link태그를 추가할때 title태그 아래에 두는 것이 익숙해서 이렇게 작성했기에 특별한 의도가 있었던 것은 아닙니다!
2. 뷰포트 너비를 디바이스 화면 너비와 동일하게 설정하고, 페이지의 기본 확대 배율을 1로 설정할 수 있습니다. 모바일에서 너무 작게보이거나 크게 보이는 것을 방지합니다.

LAH1203 replied:
지금 같은 작은 규모에선 큰 문제가 없겠지만, 보통은 meta 태그를 head 가장 앞에 배치하는 게 일반적입니다 :)
그 이유는 직접 공부하시면서 찾아보셔도 좋을 것 같아요!

LAH1203 commented on index.html:
```diff
@@ -1,13 +1,13 @@
   <body>
-    <div id="app"></div>
-    <script type="module" src="/src/main.js"></script>
+    <script type="module" src="/src/App.js" defer></script>
```
defer를 추가해주셨네요. 이유가 있을까요?

kimyou1102 replied:
미션을 진행하는 동안 요소를 제대로 잡지 못하는 문제가 발생했습니다. 이전에 비슷한 문제를 defer를 추가해서 해결한 경험이 있어서 이번에도 적용해봤어요. 하지만 지금처럼 JavaScript가 모든 요소를 렌더링하는 상황에서는 defer가 불필요하다는 걸 놓쳤네요. 미션을 빠르게 진행하다 보니 제거하지 않고 그대로 두게 된 것 같습니다..!

LAH1203 commented on style.css:
```diff
@@ -0,0 +1,76 @@
+  height: 44px;
+  min-width: 125px;
+
+  border: 1px solid #d0d5dd;
```
상수화되지 않은 일부 색상이 있는 것 같은데, 따로 기준이 있었을까요?

kimyou1102 replied:
미션에서 기본적으로 제공되는 CSS가 있어서 스타일을 분리해서 사용했습니다. 그래서 주로 CSS 파일을 어떻게 나눌지만 고민했는데 상수화에 대해서도 고민해보면 좋을 거 같네요!

LAH1203 commented on src/components/restaurantListSection/restaurantListItem/RestaurantListItem.js:
```diff
@@ -0,0 +1,48 @@
+    $distance.className = "restaurant__distance text-body";
+    $distance.textContent = `캠퍼스부터 ${distance}분 내`;
+
+    const $description = document.createElement("p");
```
createElement를 사용하면서 가독성 있게 딱딱 나눠주셨네요! 넘넘 조아요 ❤️

LAH1203 commented on src/components/restaurantFormSection/restaurantForm/RestaurantForm.js:
```diff
@@ -0,0 +1,107 @@
+import RestaurantListItem from "../../restaurantListSection/restaurantListItem/RestaurantListItem.js";
+import CategorySelect from "../categorySelect/CategorySelect.js";
+import DistanceSelect from "../distanceSelect/DistanceSelect.js";
+import "./restaurantForm.css";
```
css 파일을 html에서 link로 넣는 것과, css에서 `@import로 넣는 것과, js에서 import로 넣는 것의 차이는 뭘까요?

kimyou1102 replied:
- html에서 link를 넣는 것: 병렬방식으로 다운이 가능하며 로딩속도가 빠릅니다.
- css에서 `@import로 넣는 것: 하나의 CSS 파일을 여러군데에서 사용가능하며 로딩속도가 link보다 상대적으로 느려집니다.
- js에서 import로 넣는 것: 동적으로 CSS를 불러올 수 있지만 js가 실행되기 전에는 스타일이 적용되지 않습니다.

차이를 정확히는 몰랐는데 하리 덕분에 공부하는 기회가 되었네요😊

LAH1203 replied:
좀 더 깊이 들어가보자면, 각 방식은 스타일 코드가 언제 파싱되고 불러와지는지에 대한 시점과 다른 코드를 블로킹하지는 않는지 등에 대한 관점으로 공부해보셔도 좋을 것 같습니다~!

LAH1203 commented on src/components/restaurantFormSection/restaurantForm/RestaurantForm.js:
```diff
@@ -0,0 +1,107 @@
+
+    const formQuery = this.#getFormQuery();
+
+    const newRestaurantInfo = Object.entries(formQuery).reduce(
+      (acc, [key, query]) => {
+        acc[key] = query.value;
+        return acc;
+      },
+      {}
+    );
```
요런 로직은 유틸리티화해도 괜찮을 것 같네요~!

LAH1203 replied:
@kimyou1102 각 코멘트로 어떤 부분을 수정하셨는지는 커밋 번호와 함께 어떻게 수정하셨는지 reply로 남겨주시면 좋을 것 같습니다 :-) 2단계에서 참고해주세요~
일일이 수정 커밋을 찾아보지는 않아서요! 어떻게 반영해주셨는지를 확인하기도 어렵고요..

kimyou1102 replied:
넵 알겠습니다!
수정 커밋: [feat: 객체의 특정 기 값을 추춘하는 유틸 함수 추가](https://github.com/woowacourse/javascript-lunch/pull/189/commits/97c6459f8c72befaf5e6d03f56c8c495f4b55e0c)

LAH1203 commented on src/components/restaurantFormSection/restaurantForm/RestaurantForm.js:
```diff
@@ -0,0 +1,107 @@
+    $buttonContainer.appendChild($cancelButton);
+    $buttonContainer.appendChild($addButton);
+
+    $cancelButton.addEventListener(
+      EVENT_TYPES.click,
+      this.#handleCancelButtonClick.bind(this)
+    );
+
+    $form.addEventListener(EVENT_TYPES.submit, this.#handleSubmit.bind(this));
```
이벤트 핸들러를 등록했을 때 해제는 고려하지 않아도 괜찮을까요?

kimyou1102 replied:
render시에 새롭게 생기는 form요소에 이벤트 핸들러를 등록하기 때문에 이벤트 해제를 고려하지 않아도 된다고 생각했습니다! 또한 현재 구조상  RestaurantForm 컴포넌트는 App컴포넌트의 `initElement` 메서드에서 한번만 렌더되기에 해제하는 코드를 작성하지 않았습니다

LAH1203 replied:
넵 저는 이번 미션에 한정이 아니라 그냥 전체적으로 이벤트 해제에 대해 인지하고 고민을 해보셨으면 좋겠어서 말씀드렸습니다.
말씀해주신대로 지금 폼은 초기에 렌더링되고 이후에 돔에서 사라지거나 하지 않아 이벤트 해제를 신경쓰지 않아도 문제가 없습니다. 모달도 마찬가지고요.
근데 저희가 만약 여기서 동적인 요소를 하나둘씩 추가한다고 했을 때, 지금 모달처럼 계속 돔에 추가해놓고 css로만 사용자에게 보일지 말지 여부를 제어하는 건 바람직한 부분일까요? 돔에 보여지지 않는 요소가 너무 많아 무거워지지는 않을까요?
하지만 동적으로 컴포넌트를 추가하고 제거하더라도 대부분 이벤트 해제까지는 고려하지 않아도 동작 자체는 괜찮을 수 있습니다. 단일 흐름으로 예상하는 동작만 수행하도록 작성하면요! 그래도 예외를 배제할 수는 없고, 또 사용하지 않는 이벤트 참조로 인해 메모리 누수가 발생할 수도 있는데, 그냥 그런 다양한 관점에서 이벤트 해제에 대해 공부해보고 고민해보셨으면 좋겠습니다. 다른 크루분들과 의견을 나눠보셔도 좋고요!

LAH1203 commented on src/components/common/selectBox/SelectBox.js:
```diff
@@ -0,0 +1,42 @@
+      $option.value = option;
+
+      if (this.label === "distance") $option.textContent = `${option}분 내`;
+      else $option.textContent = option;
```
👮

kimyou1102 replied:
이 이모지는 버그를 발견했다는 의미일까요...?😱

LAH1203 replied:
아뇨! else를 사용하지 않는 게 모든 미션 공통 컨벤션으로 알고 있는데, 사용해주셔서요!

kimyou1102 replied:
헉 공통 피드백을 고려하지 못했네요😱 다음 미션부터 반영하겠습니다!!

LAH1203 reviewed (changes requested):
안녕하세요 데이지~!~! 만나서 반가워요 😆

전체적으로 코드를 일관성 있게 깔끔하게 나눠서 잘 작성해주셨더라고요! 정갈한 한식을 먹는 느낌이었달까.. 좋았씁니다 크크

고민하셨던 부분도 마지막 코멘트에서 제 나름대로 답변을 적어봤는데, 쭉 확인해보시고 수정하실 부분 있음 수정해주셔도 좋고 아니면 바로 리뷰요청 주셔도 괜찮습니당! 🤗

LAH1203 commented:
> ButtonGroup을 하나의 컴포넌트로 묶으려고 했는데 명세가 변경될 때마다 `ButtonGroup`의 `props`가 계속 늘어나게 될 가능성이 있습니다. (예를들어 type이 다르다던가, style이 달라서 className을 추가해야 한다던가 등의 추가적인 상황을 생각했습니다.)
> 그렇게 `props` 가 많아지면 코드가 복잡해지는 거 같아서 ButtonGroup을 묶지 않는 것을 선택했습니다. 재사용성과 유지보수성을 고려하면서도 이 두 개의 버튼을 효과적으로 묶을 방법이 있을지 고민입니다. 현재 방식이 괜찮을지 혹은 더 좋은 방법이 있다면 말씀해주시면 감사하겠습니다!

좀 더 컴포넌트의 목적을 생각해보시면 좋을 것 같아요! ButtonGroup의 역할은 뭘까요? 저는 버튼을 묶어주는 역할이라고 생각했어요. 그럼 버튼을 만드는 건 ButtonGroup에서 하지 않아도 되는 역할이 아닐까요?!
이 관점에서 접근하면 ButtonGroup은 버튼 배열을 전달 받기만 하면 알아서 버튼 컨테이너를 만들고, 그 안에 전달 받은 버튼을 나열해주는 동작을 해도 좋을 것 같아요!

```js
export default function ButtonGroup(buttons = []) {
  const $buttonContainer = document.createElement("div");
  $buttonContainer.className = "button-container";

  buttons.forEach($button => {
    $buttonContainer.appendChild($button);
  });

  return $buttonContainer;
}
```

거기에 더해서, 버튼 배열을 받을 때 외부에서 생성해서 전달하는 형식이 아니라 생성할 버튼에 맞는 prop 객체를 전달해서, ButtonGroup에서는 버튼을 생성해주는 방식도 괜찮을 것 같아요! 아래 참고해주세용 :-)

```js
const $buttonGroup = ButtonGroup([
  { text: "취소하기", action: "cancel", onClick: () => console.log("취소") },
  { text: "추가하기", action: "add", type: "submit", onClick: () => console.log("추가") },
]);
```

그냥 메모장에 적어봐서 제대로 동작을 할지는 모르겠는데, 어쨌든 이런 느낌으로 구현하면 컴포넌트의 역할을 좀 더 잘 나타내면서 데이지가 말씀해주신 props를 많이 추가하지 않아도 되는 확장성 있는 형태가 되지 않을까 싶어요 ㅎㅎ

> `SelectBox`를 활용한 DistanceSelect 컴포넌트를 만들었는데 추상화 레벨을 맞추려다 보니 고민이 생겼습니다. 외부에서 보면 추상화 레벨이 통일되어 가독성이 좋지만 내부에서는 추가적인 경로를 거치는 느낌이 들어서 불필요한 추상화 같기도 합니다. 추상화 레벨 맞추기 vs 접근성을 높이기 중 어떤 것을 더 신경 써야 할까요? 혹은 둘 다 만족할 수 있는 방법이 있을까요??

일단 적어주신 DistanceSelect는 불필요하다까지는 아니지만 과도한 추상화라고 생각합니다 🥲 단순히 SelectBox에 label과 options만 전달하는 역할밖에 없다면 DistanceSelect의 필요성이 애매해지는 것 같아요. 결국 코드를 보면서도 DistanceSelect는 SelectBox 인스턴스를 만들어서 반환하는 역할밖에 하는 게 없는데 왜 클래스화를 한 건지 의문이 생길 수밖에 없거든요! (CategorySelect도 마찬가지고요)
그와 별개로 `추상화 레벨 맞추기 vs 접근성을 높이기`에 대해서는 답은 없다고 생각합니당. 추상화를 한다면 그 객체를 추상화함으로써 얻는 가치가 확실해야 하고, 반대로 접근성을 높인다면 그로 인해 코드의 복잡성이 해소될 수 있을지를 봐야할 것 같아요. 근데 이건 어떤 로직이느냐에 따라 달라지는 거라... 정해진 답이 없어서 그때그때 판단은 본인의 몫입니다...ㅎㅎ "어떤 방식으로 했을 때 코드가 더 직관적인가" 이 기준으로 판단해주셔도 좋지 않을까 싶어요!

> Cypress에서 When-Then-Given 같은 가독성을 높이는 작성 방식이 있을까요? 현재 저희는 주석 없이 공통적인 부분을 함수로 추출한 후 그다음 시나리오를 테스트하는 방식으로 진행하고 있습니다. 이보다 더 가독성을 높일 방법이 있을지 하리의 의견이 궁금합니다!

테스트의 가독성을 가장 좋게 만드는 방법은 테스트에서 한 가지 동작만을 테스트하는 것이 중요하다고 생각합니다~! 예를 들면 `AddNewRestaurant.cy.js`를 보면 '음식점 추가 버튼을 누르면 바텀 시트가 열리고, 새로운 음식점 정보를 입력하여 제출 시 목록에 해당 음식점 정보가 추가된다.', '음식점 추가 버튼을 누르면 바텀 시트가 열리고, 필수 입력 정보를 입력하지 않을 시 모달창이 닫히지 않고 해당 음식점이 추가가 되지 않는다.' 같이 여러 개의 동작을 하나의 케이스에서 테스트하고 있어요! 이러면 테스트의 목적성이 떨어진다고 생각합니다.
가독성을 높이는 작성 방식은 나중에 **DCI 패턴**을 찾아보셔도 좋을 것 같아요 ㅎㅎ 저는 cypress는 아니었지만, playwright, vitest, jest에서는 잘 썼습니당. 사실 요런 패턴은 테스트 도구에는 구애받지 않는 것 같아요.

> E2E 테스트를 어디까지 커버해야 할지 고민입니다. 현재는 버튼을 클릭 → 모달오픈 → 폼 작성 → 리스트에 추가 확인까지 테스트했습니다. 한과정을 끝까지 테스트 하는 것이라 생각되어 이렇게 작성했는데 여러 개 제출을 확인하는 등 더 세부적인 테스트도 포함해야 할지 고민됩니다. E2E 테스트의 범위를 어떻게 설정하는 것이 효과적일지 궁금합니다!

일단 말씀해주신대로 핵심 동작을 E2E 테스트로 작성하는 게 가장 중요합니다~! E2E 테스트는 어느 정도 QA의 역할을 해야한다고 생각해요. 그렇지 않으면 단위 테스트랑 달리 굳이 작성하는 이유가 없어지는 것 같아요. E2E 테스트가 사용자의 플로우대로 수행했을 때 원하는대로 동작하느냐를 테스트하기 위함이지, 로직이 정상적으로 결과값을 내뱉느냐의 테스트 목적은 아니니까요!
그래서 말씀해주신 '여러 개 제출'이라는 행위 자체가 검증해야 하는 사용자의 핵심 플로우인지를 먼저 생각해보시면 좋을 것 같습니다~ 데이지가 생각하시기에 중요한 플로우라 테스트가 필요할 것 같으면 추가하시면 되고, 아니면 지금처럼 단일 아이템 추가만 확인해도 괜찮을 것 같아요!

> 컴포넌트의 응집성을 높이기 위해 각 컴포넌트마다 개별 폴더를 만들었습니다. 저희는 통일성을 위해 CSS 파일이 없는 컴포넌트도 개별 폴더를 두었는데 보다 보니 통일성은 유지되지만 오히려 가독성이 떨어지는 느낌이 듭니다. CSS가 없는 컴포넌트도 폴더를 유지하는 것이 좋을지 혹은 더 나은 구조가 있을지 하리의 의견이 궁금합니다!

사실 이런 폴더 구조는 일관성을 유지하는 게 가장 좋습니다. 프로젝트를 볼 때 가장 많이 보는 게 폴더 구조인데, 이게 나는 내 나름의 기준으로 정리했지만 다른 사람이 봤을 때는 단박에 이해되지 않는 기준으로 정리되면 보는 사람 입장에서도 파일을 찾기가 힘든 경우가 왕왕 있거든요. (제가 봤던 어떤 코드는 `components > {페이지 이름} > components, utils` 와 `components > {컴포넌트 이름}`이 뒤죽박죽 섞여 있던 형태도 있었습니다 😭)
그래서 css가 없더라도 일관성 있게 폴더를 만들어주신 건 전 오히려 가독성을 올려주는 역할도 하고 있다고 생각해요!!!!

LAH1203 commented:
추가로, UX 적으로 고려해주셔도 좋을 것 같은 부분이 있어 적어봤어요~!

1. 라벨을 클릭했을 때 input에 포커스가 가도록 하면 좋을 것 같아요!
2. 다른 input에서는 엔터를 눌렀을 때 다음 input으로 포커스가 넘어가고, 참고 링크에서만 엔터를 눌렀을 때 추가가 되는 UX는 어떨까요? 이건 호불호가 있을 수도 있어서, 그냥 적어봤어용
3. 필수 입력 데이터를 모두 입력하기 전까지는 추가하기 버튼을 disabled 처리해서 그에 맞는 스타일을 추가하면 어떨까요?!
4. 참고 링크에 링크가 아닌 다른 게 입력되었을 때, 이를 알려줘도 좋지 않을까요?

kimyou1102 committed:
refactor: 라벨 클릭시 인풋애 포커스 가도록 변경

kimyou1102 committed:
refactor: 필수 인풋 입력하지 않으면 버튼을 disabled 처리하여 제출하지 못하도록 변경

- RestaurantForm 컴포넌트는 한번만 그려지므로 addList 메서드에 버튼 disabled속성을 초기화 해주는 메서드 추가

kimyou1102 committed:
refactor: 링크 입력 밸리데이션 추가

kimyou1102 committed:
feat: 객체의 특정 기 값을 추춘하는 유틸 함수 추가

kimyou1102 committed:
refactor: quey의 value를 추출하는 부분을 extractValuesByKey 유틸함수가 하도록 변경

kimyou1102 committed:
test: 링크 밸리데이션 추가한 것에 맞게 테스트용 링크 입력을 올바른 링크 입력으로 변경

kimyou1102 committed:
test: 필수 입력 정보 미입력 시 추가하기 버튼 비활성화 테스트 수정

- 필수 입력 정보 미입력시 버튼을 비활성화하도록 바꾼 것에 맞춰 테스트도 수정

kimyou1102 committed:
refactor: form 태그에 등록된 이벤트를 change → input으로 변경

- change 이벤트는 입력 필드가 blur될 때만 발생하여 즉각적인 검사가 어려움
- 입력 즉시 유효성 검사가 이루어지도록 input 이벤트를 사용하도록 수정

kimyou1102 committed:
test: 링크 유효성 검사 테스트 추가

kimyou1102 committed:
refactor: 중복되는 로직을 가진 메서드 제거

kimyou1102 committed:
refactor: UI로직과 도메인 로직 분리를 위한 domain 클래스 생성

- addList 대신 하나의 메서드로 처리하기 위해 updateList로 변경

kimyou1102 commented:
안녕하세요 하리! 하리 덕분에 제가 짠 코드들을 다시 돌아볼 수 있는 시간이 되어서 좋았습니다. 또한 제가 고민했던 부분들에 대해서도 하리의 생각을 들어보며 다양한 시각을 얻을 수 있어서 좋았습니다🙌  

그리고 하리의 요구사항을 반영하며 2단계 요구사항 중 **도메인 영역을 TypeScript으로 변환한다**는 요구를 만족시키기 위해, 도메인과 UI 로직을 분리하는 리팩토링을 진행했습니다.  

이를 위해 `RestaurantListModal`이라는 클래스를 추가하여 식당 목록 데이터를 관리하도록 했고 `App` 컴포넌트에서는 도메인 클래스를 조작하는 메서드를 만들어 UI 로직과 분리했습니다.  

서비스 계층을 추가할지도 고민했지만 **현재 구조에서는 서비스 계층이 단순한 CRUD를 한 번 더 감싸는 역할만 하게 되어 의미가 없다고 판단했습니다.** 비즈니스 로직이 더 많아지면 그때 서비스 계층을 추가하는 것이 더 적절하다고 생각했습니다.
제 기준으로는 도메인과 UI를 적절히 분리했다고 생각하지만, 더 나은 방법이 있을 수도 있을 것 같아 의견 남겨주시면 감사하겠습니다!!

### UX 개선 반영

- 인풋 포커스 커밋: [refactor: 라벨 클릭시 인풋애 포커스 가도록 변경](https://github.com/woowacourse/javascript-lunch/pull/189/commits/dec0d5f18f57eb8cd78498b44fcf0692545b43c3)
- 버튼 disabled 처리: [refactor: 필수 인풋 입력하지 않으면 버튼을 disabled 처리하여 제출하지 못하도록 변경](https://github.com/woowacourse/javascript-lunch/pull/189/commits/d3abb4dcd00e8e6b65df250e65f3f92e0609d3b5)
- 링크 입력 밸리데이션: [refactor: 링크 입력 밸리데이션 추가](https://github.com/woowacourse/javascript-lunch/pull/189/commits/939f57dadb736ecbc5d0feb024f50086f036aefa)

LAH1203 reviewed (approved):
이미 이전 구조에 대해 리뷰 드렸었기에, 새로 만들어주신 구조에 대해 파악하고 피드백을 드리는 건 아예 별개의 새로운 리뷰가 될 것 같아 따로 피드백 드리지는 않겠습니다~! 제가 수정 요청을 드린 부분이 아니기도 해서요... 나중에 리뷰가 필요하시면 2단계 PR에서 어떤 의도로 분리하셨는지 코드와 같이 설명을 남겨주시면 좋을 것 같습니다 :-)

고생하셨습니다 데이지~ 🌼
2단계도 화이팅!
